Spigot-Server-Patches/0167-Anti-Xray.patch

```diff
@@ -0,0 +1,997 @@
++        if (x < 0) {
++            chunk = nearbyChunks[0];
++        } else if (x > 15) {
++            chunk = nearbyChunks[1];
++        } else if (z < 0) {
++            chunk = nearbyChunks[2];
++        } else if (z > 15) {
++            chunk = nearbyChunks[3];
++        }
++
++        int blockY = (chunkY << 4) + y;
++
++        if (blockY >= 0 && blockY >> 4 < chunk.getSections().length) {
++            ChunkSection chunkSection = chunk.getSections()[blockY >> 4];
++
++            if (chunkSection != Chunk.a) {
++                return chunkSection.getType(x & 15, y & 15, z & 15);
++            }
++        }
++
++        return Blocks.AIR.getBlockData();
++    }
++
++    private static boolean isSolidBlock(Block block)
++    {
++        return block.isOccluding(block.getBlockData()) && block != Blocks.MOB_SPAWNER && block != Blocks.BARRIER;
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/antixray/Lock.java b/src/main/java/com/destroystokyo/paper/antixray/Lock.java
+new file mode 100644
+index 0000000..64759ea
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/antixray/Lock.java
+@@ -0,0 +1,26 @@
++package com.destroystokyo.paper.antixray;
++
++import java.util.concurrent.atomic.AtomicInteger;
++
++public class Lock {
++    private final AtomicInteger lockCount = new AtomicInteger();
++
++    public void lock() {
++        lockCount.incrementAndGet();
++    }
++
++    public synchronized void unlock() {
++        lockCount.decrementAndGet();
++        notifyAll();
++    }
++
++    public synchronized void waitUntilUnlock() {
++        try {
++            while (lockCount.get() > 0) {
++                wait();
++            }
++        } catch (InterruptedException e) {
++            Thread.currentThread().interrupt();
++        }
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/antixray/ObfuscatorRunnable.java b/src/main/java/com/destroystokyo/paper/antixray/ObfuscatorRunnable.java
+new file mode 100644
+index 0000000..e6bdaa7
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/antixray/ObfuscatorRunnable.java
+@@ -0,0 +1,53 @@
++package com.destroystokyo.paper.antixray;
++
++import net.minecraft.server.Chunk;
++import net.minecraft.server.PacketDataSerializer;
++import net.minecraft.server.PacketPlayOutMapChunk;
++
++public class ObfuscatorRunnable implements Runnable {
++    private final PacketPlayOutMapChunk packetPlayOutMapChunk;
++    private final PacketDataSerializer packetDataSerializer;
++    private final Chunk chunk;
++    private final boolean writeSkyLightArray;
++    private final int chunkSectionSelector;
++	private final Chunk[] nearbyChunks;
++
++    public ObfuscatorRunnable(PacketPlayOutMapChunk packetPlayOutMapChunk, PacketDataSerializer packetDataSerializer, Chunk chunk, boolean writeSkyLightArray, int chunkSectionSelector, Chunk[] nearbyChunks) {
++        chunk.blocksLock.lock();
++        chunk.dataLock.lock();
++
++		if (nearbyChunks != null) {
```

Comment:
I don't know how that happened. I think I used another editor. I will fix that later.